File: tests/test_dice_policy.py
from core.dice_policy import (
    required_outcomes,
    required_shakes,
    validate_outcome_count,
)


if required_outcomes(12) != 50:
    raise SystemExit("FAIL: 12-word outcome count.")

if required_outcomes(24) != 100:
    raise SystemExit("FAIL: 24-word outcome count.")

if required_shakes(12) != 10:
    raise SystemExit("FAIL: 12-word shake count.")

if required_shakes(24) != 20:
    raise SystemExit("FAIL: 24-word shake count.")


validate_outcome_count("1" * 50, 12)
validate_outcome_count("1" * 100, 24)


try:
    validate_outcome_count("1" * 99, 24)
except ValueError:
    pass
else:
    raise SystemExit("FAIL: 99 outcomes should be rejected in production 24-word mode.")


print("12-word mode: 10 shakes / 50 outcomes")
print("24-word mode: 20 shakes / 100 outcomes")
print()
print("PASS: Dice policy tests passed.")

File: firmware/host_tests/test_dice_policy.cpp
#include "dice_policy.h"

#include <cassert>
#include <iostream>
#include <stdexcept>
#include <string>

using namespace cryptomachine;


int main() {
    assert(required_outcomes(12) == 50);
    assert(required_outcomes(24) == 100);

    assert(required_shakes(12) == 10);
    assert(required_shakes(24) == 20);

    assert(valid_dice_char('1'));
    assert(valid_dice_char('6'));

    assert(!valid_dice_char('0'));
    assert(!valid_dice_char('7'));
    assert(!valid_dice_char('x'));

    const std::string dice_12(
        50,
        '1'
    );

    const std::string dice_24(
        100,
        '6'
    );

    assert(
        validate_outcome_count(
            dice_12,
            12
        )
    );

    assert(
        validate_outcome_count(
            dice_24,
            24
        )
    );

    assert(
        !validate_outcome_count(
            dice_12.substr(0, 49),
            12
        )
    );

    assert(
        !validate_outcome_count(
            dice_24.substr(0, 99),
            24
        )
    );

    std::string invalid = dice_12;
    invalid[25] = '7';

    assert(
        !validate_outcome_count(
            invalid,
            12
        )
    );

    bool rejected = false;

    try {
        required_outcomes(18);
    }
    catch (const std::invalid_argument&) {
        rejected = true;
    }

    assert(rejected);

    std::cout
        << "PASS: C++ V1 dice policy tests"
        << std::endl;

    return 0;
}
